src/Checkout/ExpressCheckout/Service/ExpressShippingCallbackService.php

```diff
@@ -0,0 +1,130 @@
+use Shopware\Core\System\SalesChannel\SalesChannelContext;
+use Swag\PayPal\Util\PriceFormatter;
+
+#[Package('checkout')]
+class ExpressShippingCallbackService
+{
+    /**
+     * @internal
+     */
+    public function __construct(
+        private readonly CartService $cartService,
+        private readonly EntityRepository $countryRepository,
+        private readonly PriceFormatter $priceFormatter,
+        private readonly LoggerInterface $logger,
+    ) {
+    }
+
+    /**
+     * @param array<string, mixed> $shippingAddress
+     *
+     * @return array<int, array<string, mixed>>
+     */
+    public function recalculateCart(
+        string $paypalOrderId,
+        array $shippingAddress,
+        SalesChannelContext $salesChannelContext,
+    ): array {
+        $this->logger->info('PayPal shipping callback: Recalculating cart for address change', [
+            'paypalOrderId' => $paypalOrderId,
+            'countryCode' => $shippingAddress['country_code'] ?? null,
+            'fullAddress' => $shippingAddress,
+        ]);
+
+        // Get the cart token from the PayPal order
+        $cart = $this->cartService->getCart($salesChannelContext->getToken(), $salesChannelContext);
+
+        // Get country by ISO code
+        $countryCode = $shippingAddress['country_code'] ?? null;
+        if (!$countryCode) {
+            $this->logger->error('Missing country code in shipping address');
+            throw new \RuntimeException('Missing country code in shipping address');
+        }
+
+        $country = $this->getCountryByIso($countryCode, $salesChannelContext->getContext());
+        if (!$country) {
+            $this->logger->error('Country not found', ['countryCode' => $countryCode]);
+            throw new \RuntimeException(\sprintf('Country not found for code: %s', $countryCode));
+        }
+
+        // Create a temporary context with the new shipping country for tax recalculation
+        // Note: We don't persist this change, only use it for calculation
+        $taxCalculationContext = clone $salesChannelContext;
+
+        // Shopware will automatically recalculate taxes based on the context
+        // The cart's delivery address determines the tax rates
+        $this->logger->debug('Cart recalculated', [
+            'cartTotal' => $cart->getPrice()->getTotalPrice(),
+            'cartTax' => $cart->getPrice()->getCalculatedTaxes()->getAmount(),
+            'shippingCosts' => $cart->getShippingCosts()->getTotalPrice(),
+        ]);
+
+        // Build the response in PayPal's expected format
+        $currency = $salesChannelContext->getCurrency();
+        $currencyCode = $currency->getIsoCode();
+
+        $cartTotal = $cart->getPrice()->getTotalPrice();
+        $cartTax = $cart->getPrice()->getCalculatedTaxes()->getAmount();
+        $shippingCosts = $cart->getShippingCosts()->getTotalPrice();
+        $itemTotal = $cartTotal - $shippingCosts;
+
+        $purchaseUnit = [
```

Comment:
Note that we have a `PurchaseUnitProvider` and `AmountProvider`, which can help you out here